---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%",
  warnings = FALSE,
  error = FALSE,
  message = FALSE
)

```
# superintronic <img id="superintronic" src="man/figures/superintronic.png" align="right" width = "125" />

[![lifecycle](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![Build Status](https://travis-ci.org/sa-lee/superintronic.svg?branch=master)](https://travis-ci.org/sa-lee/superintronic)
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/sa-lee/superintronic?branch=master&svg=true)](https://ci.appveyor.com/project/sa-lee/superintronic)
[![Codecov test coverage](https://codecov.io/gh/sa-lee/superintronic/branch/master/graph/badge.svg)](https://codecov.io/gh/sa-lee/superintronic?branch=master)



_superintronic_ centers around exploring coverage over genomic
regions via computing simple summary statistics and visualisations. 
The aim is to provide a modular worklfow via an interface built on top of the 
[_plyranges_](https://sa-lee.github.io/plyranges/index.html) package and 
emphasising simple 


## Quick start 

Often, as quality control check in sequencing analysis you would like
to visualise coverage within each sample over the exon/intron parts of 
gene of interest.  This is simple with superintronic! 

As an example, we will use the sample
BAM files from the  [`airway`](https://bioconductor.org/packages/release/data/experiment/html/airway.html) data package in Bioconductor and our target gene is _SRM_  

The basic workflow consists of three steps  

1. Setting up your gene region of interest 
2. Computing coverage
3. Visualising coverage results

(_Note that we are considering a single gene for illustrative purposes
and in principle you can look at coverage over any genomic region you're 
interested in!_)

### 1. Setting up annotation GRanges

You can start from a GTF/GFF file for your given organism and `collect_parts()`
to gather up the intronic/exonic regions for each gene. You can then use the
_plyranges_ grammar to select your target gene.

```{r parts, eval = TRUE}
library(superintronic)
suppressPackageStartupMessages(library(plyranges))

features <- system.file("extdata", 
                        "Homo_sapiens.GRCh37.75_subset.gtf", 
                        package = "airway") %>% 
  collect_parts() %>% 
  filter(gene_name == "SRM")

features
```


### 2. Computing coverage over samples

You can compute a long form GRanges containing coverage scores in parallel over
a set of BAM files. All you need to begin is a character vector containing
the path the BAM files or a `data.frame` representing the experimental design
that contains a column of BAM files.

```{r design, eval = TRUE}
design <- read.csv(system.file("extdata", 
                               "sample_table.csv", 
                               package = "airway")) %>% 
  dplyr::select(sample_id = X, cell, dex, albut) %>% 
  dplyr::mutate(bam = dir(system.file("extdata", package = "airway"), 
                          pattern = "*.bam",
                          full.names = TRUE)
  )

cvg <- compute_coverage_long(design, source = "bam")

cvg 
```

This function automatically propagates, the metadata associated
with a design onto the resulting GRanges. You can speed up this computation
if your BAM files are indexed by providing a target GRanges so coverage 
will be restricted to that set of ranges only.

Once the coverage has been computed as a GRanges object, 
we can then intercet overlapping gene parts (from `collect_parts()`) 

```{r join, eval = TRUE}
cvg_over_features <- cvg %>% 
  select(-bam) %>% 
  join_parts(features)
cvg_over_features
```

Further summaries can then be computed using rangewise diagnostics,
see the `rangle()` function for details.

### 3. Visualising coverage scores

Coverage over a gene or a genomic region can be constructed via
the `view_coverage()` function. The coverage score is disjoined
and aggregated over all samples that overlaps the target region:

```{r cov-01, eval = TRUE}
view_coverage(cvg_over_features, features)
```

By default, the coverage is aligned from the 5' to the 3' end of the target
region, and exon regions are coloured green and intron parts are colour orange. 
The function returns a ggplot object, so extra annotations or layers can be added
via the `ggplot2` library. 


Coverage scores can also be facetted if additional 
experimental design information has been added. For example, we can split the
coverage by each sample or treatment group or cell type


```{r cov-02}
view_coverage(cvg_over_features, features, facets = "cell")
```


## Learning more

_superintronic_ provides a general API to compute not only coverage, 
but also aggregate and visualise other measurements of interest over
any arbirtrary genomic region. See the [vignette]() for details and 
the [_plyranges_](https://sa-lee.github.io/plyranges/index.html#learning-more)
documentation. 