---
title: "Exploring intron signal with superintronic"
author: "Stuart Lee, Charity Law"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
                      fig.align = "center",
                      comment = "#>", 
                      message = FALSE,
                      warning = FALSE
                      )

suppressPackageStartupMessages(library(plyranges))

# pre cached results
cvg_fn <- here::here("data", "cvg.rds")
skip <- file.exists(cvg_fn)
if (skip) {
  cvg_over_features <- readr::read_rds(cvg_fn)
}

```

## Preparing GFF/GTF files

```{r gff}
library(superintronic)
gff <- here::here("data-raw", "gencode.v27.annotation.gtf.gz")
ref <- BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38 %>% 
  get_genome_info() %>% 
  GenomeInfoDb::keepStandardChromosomes("Homo sapiens", "coarse")

gr_gff <- read_gff(gff, genome_info = ref )
anno <- prepare_annotation(gr_gff)
anno 
```

Once the annotation has been supplied you can explore various aspects,
```{r anno-summary}
anno %>% 
  group_by(n_olaps) %>% 
  summarise(n = n())
```

and given a set of rules, filter the annotation (by default we include
protein coding, genes with more than one exon, genes that do not overlap
any other genes)

```{r default-filter}
anno_sub <- anno %>% 
  filter_rules()

anno_sub
```


## Preparing BAM files

Generate a design data.frame containing filenames, and other variables.

```{r prepare-design}
tbl <- read.table(
  here::here("data-raw", "targets.txt"),
  header = TRUE,
  stringsAsFactors = FALSE
  ) %>% 
  transform(File = S4Vectors::Rle(sub("\\.", "-", File))) %>%
  S4Vectors::DataFrame()

design <- tbl %>% 
  BiocGenerics::subset(Replicate %in% c("R1", "R2", "R3") & Mixture == 0) %>%  
  S4Vectors::transform(Sample = paste0(Replicate, "_", "HCC287"),
            CellLine = "HCC287")
design
```


Now we can take our BAM files and compute coverage. This
produces a large GRanges object, with two metadata columns,
`source` which is the BAM file  that coverage was computed for, and `score`
which is the coverage score. By default, this function computes
the coverage scores in parallel using BiocParallel.

An optional GRanges object may be passed to this function that will 
restrict the coverage computation to those regions; and set the annotation
in the resulting GRanges object. 

```{r compute-coverage, eval = skip}
cvg <- gather_coverage(here::here("data-raw", design$File),
                       genome_info = ref)

```

## Coverage over intronic/exonic regions

Now all the ingredients are in place to merge the coverage scores
to our prepared annotation with `merge_coverage()`. This again
returns a GRanges object, restricted to the intersection of the 
coverage ranges with intron/exon ranges. Additional columns are added
corresponding to the properties of the intron/exon.

```{r merge-annotations, eval = skip}
cvg_over_features <- merge_coverage(cvg, anno_sub)

cvg_over_features
```

We can also add in elements of the experimental design by joining the design
table to the coverage ranges:

```{r merge-design, eval = skip}
per_replicate_avg <- cvg %>% 
  group_by(source) %>% 
  summarise(average_replicate_coverage = mean_coverage(score, width)) 

per_replicate_avg <- per_replicate_avg %>% 
  base::merge(design[, c("Replicate", "Kit", "File")], 
        by.x = "source", 
        by.y = "File") 
  
per_replicate_avg <- as(per_replicate_avg, "DataFrame") %>% 
  S4Vectors::transform(source = S4Vectors::Rle(source),
                       Replicate = S4Vectors::Rle(Replicate),
                       Kit = S4Vectors::Rle(Kit))

cvg_over_features <- merge_design(cvg_over_features,
                                  per_replicate_avg,
                                  on = "source")
```


Now we will compute summaries over each Kit, in the following manner

* Calculate mean exon coverage per gene as `exon_avg`
* Calculate mean intron coverage per gene, `intron_avg`



```{r filter-new}
gene_avg <- cvg_over_features %>% 
  group_by(Kit, gene_id) %>% 
  summarise(gene_average = mean_coverage(score, width)) %>% 
  dplyr::as_tibble()

intron <- cvg_over_features %>% 
  filter(feature_type == "intron") %>% 
  group_by(Kit, gene_id) %>% 
  summarise(intron_average = mean_coverage(score, width),
            score = score, 
            n_bases = width,
            n_replicate = n_distinct(source),
            feature_length = sum(unique(feature_length))) %>% 
  dplyr::as_tibble() 

exon <- cvg_over_features %>% 
  filter(feature_type == "exon") %>% 
  group_by(Kit, gene_id) %>% 
  summarise(exon_average = mean_coverage(score, width)) %>% 
  dplyr::as_tibble() %>% 
  group_by(Kit) %>% 
  mutate(E = mean(exon_average))

all_features <- dplyr::left_join(
  gene_avg,
  intron %>% dplyr::left_join(exon)
) %>% 
  select(Kit, gene_id, dplyr::ends_with("average"), dplyr::everything())
all_features <- all_features %>% 
  group_by(Kit) %>% 
  mutate(
    total_bases_above_E = sum(as(n_bases, "IntegerList")[as(score, "IntegerList") > E]))
all_features
```


Now we can try out some filters...

```{r}
hits <- all_features %>% 
  group_by(Kit) %>% 
  filter(exon_average > E, intron_average < 6, total_bases_above_E > 5000) %>% 
  ungroup() %>% 
  group_by(gene_id)
hits
```

And plot the results:

```{r}
save_plot <- function(.x, .y) {
  if (nrow(.x) == 2) {
    fname <- paste0(.y$gene_id,  "-both-ir.png")
  } else {
    fname <- paste0(.y$gene_id, "-", .x$Kit, "-ir.png")
  }
  fname <- here::here("figures", "new-filter", fname)
  p <- anno_sub %>% 
    filter(gene_id == !!.y$gene_id) %>% 
    view_coverage(cvg_over_features, ., hline, "Kit")
  ggplot2::ggsave(fname, p)
} 

hline <- dplyr::distinct(all_features, Kit, E)


hits %>% 
  dplyr::group_walk(save_plot)
  

```
