---
title: "Exploring intron signal with superintronic"
author: "Stuart Lee, Charity Law"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
                      fig.align = "center",
                      comment = "#>", 
                      message = FALSE,
                      warning = FALSE
                      )
```

## Preparing GFF/GTF files

```{r gff}
gff <- here::here("data-raw", "gencode.v27.annotation.gtf.gz")
ref <- BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38 %>% 
  get_genome_info() %>% 
  GenomeInfoDb::keepStandardChromosomes("Homo sapiens", "coarse")

gr_gff <- read_gff(gff, genome_info = ref )
anno <- prepare_annotation(gr_gff)
anno 
```

Once the annotation has been supplied you can explore various aspects,
```{r anno-summary}
anno %>% 
  group_by(n_olaps) %>% 
  summarise(n = n())
```

and given a set of rules, filter the annotation (by default we include
protein coding, genes with more than one exon, genes that do not overlap
any other genes)

```{r}
anno_sub <- anno %>% 
  filter_rules()

anno_sub
```


## Preparing BAM files

Generate a design data.frame containing filenames, and other variables.

```{r prepare-design}
tbl <- read.table(
  here::here("data-raw", "targets.txt"),
  header = TRUE,
  stringsAsFactors = FALSE
  ) %>% 
  transform(File = S4Vectors::Rle(sub("\\.", "-", File))) %>%
  S4Vectors::DataFrame()

design <- tbl %>% 
  BiocGenerics::subset(Replicate %in% c("R1", "R2", "R3") & Mixture == 0) %>%  
  S4Vectors::transform(Sample = paste0(Replicate, "_", "HCC287"),
            CellLine = "HCC287")
design
```


Now we can take our BAM files and compute coverage. This
produces a large GRanges object, with two metadata columns,
`source` which is the BAM file  that coverage was computed for, and `score`
which is the coverage score. By default, this function computes
the coverage scores in parallel using BiocParallel.

An optional GRanges object may be passed to this function that will 
restrict the coverage computation to those regions; and set the annotation
in the resulting GRanges object. 

```{r compute-coverage}
cvg <- gather_coverage(here::here("data-raw", design$File),
                       genome_info = ref)
```

## Coverage over intronic/exonic regions

Now all the ingredients are in place to merge the coverage scores
to our prepared annotation with `merge_coverage()`. This again
returns a GRanges object, restricted to the intersection of the 
coverage ranges with intron/exon ranges. Additional columns are added
corresponding to the properties of the intron/exon.

```{r}
cvg_over_features <- merge_coverage(cvg, anno_sub)

cvg_over_features
```

We can also add in elements of the experimental design by joining the design
table to the coverage ranges:

```{r}

cvg_over_features <- merge_design(cvg_over_features,
                                  design,
                                  on = "File")
```


Now we can begin summarising, here we spread the coverage table into wide
form by aggregating exon/intron coverage over each gene over all samples.
We can also add a thresholding argument to compute the number of bases 
above some score threshold and the proportion of bases above that threshold
that cover the total length of an exon/intron within a gene.

```{r}

exin_by_kit <- spread_coverage(
  cvg_over_features,
  Kit,
  threshold = 50L
)

view_exin(exin_by_kit,
          x = log2(exon.average_coverage + 1),
          y = log2(intron.average_coverage + 1),
          facets = "Kit"
          ) 

view_exin(exin_by_kit,
          x = log2(exon.average_coverage + 1),
          y = log2(intron.n_bases_above_threshold + 1)  ,
          facets = "Kit"
          ) 


```


Now let's look at coverage plots for genes we designate as 'interesting'.

First we look at genes that have a large number of intron bases covered
above a score of 50 and an log2 average exon coverage above 5. 

```{r}
ir_genes <- subset(exin_by_kit, 
                   (log2(intron.n_bases_above_threshold + 1) > 10 & log2(exon.average_coverage + 1) > 5),
                   select = "gene_id"
                   ) %>% 
  unique() %>% 
  .$gene_id %>% 
  S4Vectors::runValue()
```  

This produces `r length(ir_genes)` candidates for intron retention. We can 
look at their coverage profiles as follows:

```{r, eval = FALSE}
for (gene in ir_genes) {
  fname <- here::here("figures", paste0(gene, "ir.cov.png"))
  if (file.exists(fname)) break
  p <- anno_sub %>% 
    filter(gene_id == !!gene) %>% 
    view_coverage_within_gene(cvg_over_features, ., "Kit")
  ggplot2::ggsave(fname, p)
}
```


Next we select genes that have both high exon and intron coverage and plot
their coverage profiles:

```{r}
high_intron_exon <-  subset(exin_by_kit, 
                   (log2(exon.average_coverage + 1) > 7  & log2(intron.average_coverage) > 5 ),
                   select = "gene_id"
                   ) %>% 
  unique() %>% 
  .$gene_id %>% 
  S4Vectors::runValue()

n_hie <- length(high_intron_exon)

high_intron_exon <- setdiff(high_intron_exon, ir_genes)
```

Of the `r n_hie - length(high_intron_exon)` genes not classified as intron retention , 
there were `r length(high_intron_exon)` genes that met this criteria. 

```{r, eval = FALSE}
for (gene in high_intron_exon) {
  fname <- here::here("figures", paste0(gene, "high.cov.png"))
  if (file.exists(fname)) break
  p <- anno_sub %>% 
    filter(gene_id == !!gene) %>% 
    view_coverage_within_gene(cvg_over_features, ., "Kit")
  ggplot2::ggsave(fname, p)
}

```

Finally, if we look at genes that have high exon coverage but low
intron coverage:

```{r}
low_intron_high_exon <- subset(exin_by_kit, 
                   (log2(exon.average_coverage + 1) > 8 & log2(intron.average_coverage) < 1),
                   select = "gene_id"
                   ) %>% 
  unique() %>% 
  .$gene_id %>% 
  S4Vectors::runValue()
low_intron_high_exon <- setdiff(low_intron_high_exon, c(high_intron_exon, ir_genes))
```

As expected these `r length(low_intron_high_exon)` genes don't overlap
either of the above categories. We also construct their profiles.

```{r, eval = FALSE}
for (gene in low_intron_high_exon) {
  fname <- here::here("figures", paste0(gene, "low.cov.png"))
  if (file.exists(fname)) break
  p <- anno_sub %>% 
    filter(gene_id == !!gene) %>% 
    view_coverage_within_gene(cvg_over_features, ., "Kit")
  ggplot2::ggsave(fname, p)
}
```